### PR TITLE
[core] Added implementation of a ChannelGroupUID

### PR DESCRIPTION
**Pull request by cweitkamp**
Initial idea and for discussion.

As preparation for a `ChannelGroupBuilder` which is accessible from the `ThingHandlerCallback` similar to the `createChannelBuilder()` or `editChannel()` methods I would like to introduce a `ChannelGroupUID` to identify the channel group to be created / edited.

- Added implementation and test of a `ChannelGroupUID`

Signed-off-by:Christoph Weitkamp <github@christophweitkamp.de>

### CODE REVIEW

**Comment by cweitkamp** [comment]
Yes, of course. Currently there is no easy way to create all channels for a specific `ChannelGroupType`. It has to be done manually by creating one channel after the other. I would like to introduce a new feature which can be used to create those channels all at once. An automatic approach needs access to the `ChannelGroupTypeRegistry` and/or the `ChannelTypeRegistry` which is not allowed.

I am not sure if we need a real builder for it or if we even need a `ChannelGroup` entity. We can can e.g. just return a list of `ChannelBuilders`.

```
public List<ChannelBuilder> createChannelGroupBuilder(ChannelGroupUID channelGroupUID, ChannelGroupTypeUID channelGroupTypeUID) {
    ChannelGroupType channelGroupType = channelGroupTypeRegistry.getChannelGroupType(channelGroupTypeUID);
    if (channelGroupType == null) {
        throw new IllegalArgumentException(String.format("Channel group type '%s' is not known", channelGroupTypeUID));
    }
    List<ChannelBuilder> channelBuilders = new ArrayList<>();
    for (ChannelDefinition channelDefinition : channelGroupType.getChannelDefinitions()) {
        ChannelType channelType = channelTypeRegistry.getChannelType(channelDefinition.getChannelTypeUID());
        if (channelType != null) {
            ChannelUID channelUID = new ChannelUID(channelGroupUID, channelDefinition.getId());
            channelBuilders.add(ThingFactoryHelper.createChannelBuilder(channelUID, channelType, configDescriptionRegistry));
        }
    }
    return channelBuilders;
}
```

**Comment by cweitkamp** [comment]
@htreu Since I already posted an initial code snippet I have created some working code too.

https://github.com/cweitkamp/smarthome/compare/feature-channelgroupuid...cweitkamp:feature-channelgroupbuilder

**Comment by cweitkamp** [comment]
I am with you. Unused code is just a millstone around the neck. But of course I have a use case for it.

The OWM binding (#5694) supports different forecast strategies (hourly / daily). Currently I decided to use a fixed number of hours (24) / days (5) to be requested whereas the OWM API is capable of delivering more data (max. 5 days / hour forecast and 16 days / daily forecast). Instead of defining fixed numbers I want to add configuration parameter for them. This should make it possible for everyone to adopt the number of forecasted hours / days for their own needs.

To archive my goal it would be nice to have an easy way to create a dynamical number of channel groups.

**Comment by htreu** [comment]
Yes, I just started to review the OWM binding and realised your needs in this. Thanks for the explanations.
I will merge here, so you can go on with the rest of the puzzle ;-)

File: bundles/core/org.eclipse.smarthome.core.thing.test/src/test/java/org/eclipse/smarthome/core/thing/ChannelGroupUIDTest.java [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+
+    @Test(expected = IllegalArgumentException.class)
+    public void testMultipleChannelGroupSeparators() {
+        new ChannelGroupUID("binding:thing-type:thing:group#id#what_ever");
+    }
+
+    @Test
+    public void testChannelGroupUID() {
+        ChannelGroupUID channelGroupUID = new ChannelGroupUID(THING_UID, GROUP_ID);
+        assertEquals("binding:thing-type:thing:group#", channelGroupUID.toString());
```
[the remark this answers is not in the document]
Comment (cweitkamp) [reply bundles/core/org.eclipse.smarthome.core.thing.test/src/test/java/org/eclipse/smarthome/core/thing/ChannelGroupUIDTest.java]: Okay. I prepare a follow up to remove the `#` at the end.

File: bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/ChannelGroupUID.java [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+/**
+ * {@link ChannelGroupUID} represents a unique identifier for channel groups.
+ *
+ * @author Christoph Weitkamp - Initial contribution
+ */
+@NonNullByDefault
+public class ChannelGroupUID extends UID {
+
+    private static final String CHANNEL_GROUP_SEGMENT_PATTERN = "[\\w-]*#";
+    private static final String CHANNEL_GROUP_SEPERATOR = "#";
```
[the remark this answers is not in the document]
Comment (cweitkamp) [reply bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/ChannelGroupUID.java]: Hehe, that is a little copy&paste error from `ChannelUID` :wink:. I will change it.

File: bundles/core/org.eclipse.smarthome.core.thing.test/src/test/java/org/eclipse/smarthome/core/thing/ChannelGroupUIDTest.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (cweitkamp) [reply bundles/core/org.eclipse.smarthome.core.thing.test/src/test/java/org/eclipse/smarthome/core/thing/ChannelGroupUIDTest.java]: See #6157

**Pull request merged**